polish up the noncopyable diagnostics

kavon opened the pull request:
🚧 here be lots of find-replace 🚧

- replaces "move-only" terminology with "noncopyable"
- replaces compiler jargon like "guaranteed parameters"
  and "lvalue" with corresponding language-level notions
- simplifies diagnostics about closures.

and probably more.

rdar://109281444

kavon committed:
[nfc] delete dead code

kavon committed:
polish noncopyable types diagnostic wordings

- replaces "move-only" terminology with "noncopyable"
- replaces compiler jargon like "guaranteed parameters"
  and "lvalue" with corresponding language-level notions
- simplifies diagnostics about closures.

and probably more.

rdar://109281444

kavon commented:
@swift-ci smoke test linux

kavon committed:
reword diagnostics for partial consume of fields

also moves the error to the invalid partial consume.

rdar://109281444

kavon committed:
reword "boundary use" to just "non-consuming"

rdar://109281444

kavon committed:
update diagnostics in clang importer

rdar://109281444

kavon commented:
@swift-ci please test

gottesmm commented on include/swift/AST/DiagnosticsSIL.def:
```diff
@@ -753,50 +753,32 @@ ERROR(sil_moveonlychecker_owned_value_consumed_and_used_at_same_time, none,
-      "'%0' consumed but not reinitialized before end of function", (StringRef))
-ERROR(sil_moveonlychecker_inout_not_reinitialized_before_end_of_closure, none,
-      "'%0' consumed in closure but not reinitialized before end of closure", (StringRef))
+      "missing reinitialization of '%0' after consume", (StringRef))
```
By removing the information explaining to the user that the value needs to be reinitialized before the end of the function, the error is no longer actionable for the user unless they understand the properties of inout. Can you add that information back in?

gottesmm commented on include/swift/AST/DiagnosticsSIL.def:
```diff
@@ -832,10 +811,22 @@ NOTE(sil_movekillscopyablevalue_use_here, none,
+      "usage of no-implicit-copy value that the compiler can't verify. This is a compiler bug. Please file a bug with a small example of the bug", ())
+ERROR(sil_moveonlychecker_not_understand_moveonly, none,
+      "usage of a noncopyable type that compiler can't verify. This is a compiler bug. Please file a bug with a small example of the bug"
+      "check!", ())
```
I think you left in "check!" here.

gottesmm commented on test/SILOptimizer/moveonly_objectchecker_diagnostics.swift:
```diff
@@ -3041,24 +3041,24 @@ public func closureVarCaptureClassOwnedArgUseAfterConsume3(_ x2: __owned Klass)
     f()
-    let x3 = x2 // expected-error {{'x2' was consumed but it is illegal to consume a noncopyable mutable capture of an escaping closure. One can only read from it or assign over it}}
-    consumeVal(x2) // expected-error {{'x2' was consumed but it is illegal to consume a noncopyable mutable capture of an escaping closure. One can only read from it or assign over it}}
+    let x3 = x2 // expected-error {{captured 'x2' cannot be consumed within a closure}}
```
This is an example of a case where we have a consuming use of a value outside of a closure that is banned.

gottesmm commented on include/swift/AST/DiagnosticsSIL.def:
```diff
@@ -753,50 +753,32 @@ ERROR(sil_moveonlychecker_owned_value_consumed_and_used_at_same_time, none,
-      "'%0' consumed in closure. This is illegal since if the closure is invoked more than once the binding will be uninitialized on later invocations", (StringRef))
+      "'%0' is borrowed and cannot be consumed by closure capture", (StringRef))
+
+ERROR(sil_moveonlychecker_capture_consumed_in_closure, none,
```
I looked through the patch and I think you are using this for escaping closures. I think there are two problems here:

1. I think the information that the closure is escaping is important for the user to understand. We should tell the user that especially since we made the change to make non-escaping lets non-escaping. That provides a lot of clarity to the user.
2. This diagnostic as shown below also occurs when a value that is captured by an escaping closure is used outside of the closure itself (the property is not flow sensitive). So the diagnostic says the use is in a closure, but it clearly isn't which is misleading.

kavon replied:
Seems like the main thing we need to communicate is that a mutable capture can be consumed in a closure, but an immutable one cannot. That doesn't change whether the closure is passed to an `@escaping` parameter or not.

kavon committed:
specify "inout" for missing reinit error

rdar://109281444

kavon committed:
remove "within a closure" since you can't consume any captured noncopyable binding

rdar://109281444

kavon committed:
tighten up consistency in terminology

- refer to a "consuming use" as simply a "consume", to reserve "use" for non-consuming uses.
- refer to "non-consuming uses" as just a "use".
- don't call it a "user defined deinit" and instead a "deinitializer" to match Sema
- be specific about what binding a closure is capturing that is preventing consumption.

rdar://109281444

kavon committed:
merge similar diagnostics together under a unified naming scheme for more consistent word tense

sil_movekillscopyablevalue_* and sil_moveonlychecker_* can share diagnostics.

rdar://109281444

kavon committed:
reword 'other consume here' to 'consumed again here'

this also fixes a bug where sometimes we simply emit
'consumed here' twice and other times we'd said 'other
consume here' for the same "consumed more than once"
message. so I went through and changed all of the 2nd
consumes into "consumed again".

rdar://109281444

kavon commented:
@swift-ci please test and merge
